Add Caltech101 dataset

goodhamgupta opened the pull request:
Hi everyone,

This PR aims to add support for the Caltech101 dataset(mentioned in issue #16). 

The Caltech101 dataset:
- Consists of 9145 images belonging to a total of 102 classes.
- The size of each image varies between 200-300 pixels in both width and height.

There is no test split available for the current dataset. The implementation is based loosely on the [Tensorflow Datasets Library](https://github.com/tensorflow/datasets/blob/master/tensorflow_datasets/image_classification/caltech.py). 

The dataset can be queried as follows:

```ex
{{image_list, image_type, image_shapes}, {label_list, label_type, label_shape}} = Scidata.Caltech101.download()
```

`image_shapes` consist of the shape for each image. For now, there is no way to transform the images after downloading them.

There are a few things that I wasn't able to figure out, such as:
- For downloading the dataset, the official website provides a google drive link, but I was unable to get around the "Google Virus Scan warning" redirect page for large files. Due to this, I've switched the download link to the Fast AI link hosted on S3.
- I was unable to obtain the image shape from the binary returned after extracting the image using `erl_tar`. Due to this, I've added a new dependency(evision) and use it to find the dimensions of the image.


Thanks!

josevalim commented on lib/scidata/caltech101.ex:
```diff
@@ -0,0 +1,175 @@
+    records =
+      files
+      |> Enum.map(fn {fname, _image} -> Task.async(fn -> generate_records(fname) end) end)
+      |> Enum.map(&Task.await(&1, :infinity))
```
Please use Task.async_stream. We can make the concurrency configurable perhaps?

goodhamgupta replied:
I've added the option to read concurrency from the environment variable called "MAX_CONCURRENCY". Would this work?

josevalim commented:
Thank you @goodhamgupta!

However, I don't think we should add a dependency to `evision`, as that would make this library "heavier". We have two options:

1. Make evision an optional dependency

2. I assume we can hardcode the shapes, as they are constant and known upfront, right? If so, we hardcode them and emit the images as is. The caller can then call OpenCV

WDYT?

goodhamgupta commented:
> Thank you @goodhamgupta!
> 
> However, I don't think we should add a dependency to `evision`, as that would make this library "heavier". We have two options:
> 
>     1. Make evision an optional dependency
> 
>     2. I assume we can hardcode the shapes, as they are constant and known upfront, right? If so, we hardcode them and emit the images as is. The caller can then call OpenCV
> 
> 
> WDYT?

Thanks for the quick review @josevalim! 

Regarding the `evision` library:

- I agree that adding it makes the library heavier. However, I noticed that the `IMDBReviews` class has the option to transform the inputs and labels using a custom-provided function. Do you think we should add support for the Caltech101 dataset as well, or leave this implementation to the caller?
- As you mentioned, having a file containing the shapes is easier to implement. The only case where this could break is if the source dataset was updated on the same URL. In this case, having the `evision` dependency will help, but I guess this scenario is unlikely to happen.

Tensorflow Datasets also maintains a separate file for the [labels](https://github.com/tensorflow/datasets/blob/master/tensorflow_datasets/image_classification/caltech101_labels.txt) for some datasets. I think we could also follow this approach. 

For now, I think we could just have files containing the image shapes and the labels. WDYT?

josevalim commented:
>  Do you think we should add support for the Caltech101 dataset as well, or leave this implementation to the caller?

We removed almost all transform_input and transform_labels stuff because they were redundant. So I think we should remove them from there too? 🤔 

> In this case, having the evision dependency will help, but I guess this scenario is unlikely to happen.

Yeah, my understanding is that those datasets are usually set in stone and rarely ever change.

goodhamgupta commented on lib/scidata/caltech101.ex:
```diff
@@ -0,0 +1,83 @@
+    label = Map.get(label_mapping, class_name)
+    shape = image_mapping |> Map.get("#{label}/#{image_fname}") |> List.to_tuple()
+
+    {image, label, shape}
```
For some reason, the byte_size of the binary returned after extracting the tar file using :erl_tar does not match the byte_size returned when we open the file using OpenCv(via evision) and return the binary.

Example:

- For the file `101_ObjectCategories/BACKGROUND_Google/image_0004.jpg`, the image shape is `(200,200,3)`. This record is typically the first returned when we load the dataset using:
```ex
iex> {images, labels} = Scidata.Caltech101.download()
iex> {bin_images, img_type, img_shapes} = images
iex> bin_images |> List.first |> byte_size
10071
iex> img_shape = img_shapes |> List.first
iex> bin_images |> List.first |> Nx.from_binary(img_type) |> Nx.reshape(img_shape)
** (ArgumentError) cannot reshape, current shape {10071} is not compatible with new shape {200, 200, 3}
    (nx 0.1.0-dev) lib/nx/shape.ex:159: Nx.Shape.reshape/2
    (nx 0.1.0-dev) lib/nx.ex:1818: Nx.reshape/3
```

However, reading the file using OpenCV gives the expected shape:

```ex
iex> {:ok, mat} = OpenCV.imread("101_ObjectCategories/BACKGROUND_Google/image_0004.jpg")
iex> {:ok, type} = OpenCV.Mat.type(mat)
iex> {:ok, {rows, cols, channels}} = OpenCV.Mat.shape(mat)
iex> {:ok, binary_data} = OpenCV.Mat.to_binary(mat)
iex> binary_data |> byte_size
120000
iex> binary_data |> Nx.from_binary(img_type) |> Nx.reshape(img_shape)
#Nx.Tensor<
  u8[200][200][3]
  [
    [
      [26, 49, 44],
      [34, 54, 49],
      [29, 41, 35],
      [34, 41, 34],
      [64, 64, 58],
      [64, 65, 55],
      [39, 45, 34],
      [28, 43, 29],
      [58, 83, 69],
      [40, 76, 60],
      [52, 90, 78],
      [51, 91, 80],
      [38, 72, 65],
      [52, 82, 77],
      [63, 83, 84],
      [35, 52, 49],
      [19, 36, ...],
      ...
    ],
    ...
  ]
>
```

Any hints on how to fix this issue? 😅

josevalim commented:
Yes, the original images likely have .jpeg/.png header metadata and stuff. So you will need OpenCV/Evision, I am just thinking the caller of the library would do it instead of the lib. WDYT?

goodhamgupta commented:
> Yes, the original images likely have .jpeg/.png header metadata and stuff. So you will need OpenCV/Evision, I am just thinking the caller of the library would do it instead of the lib. WDYT?

If I understand it correctly, we would then need to return the file paths for each image instead of the binary? I'm not sure how to convert the binary extracted from erl_tar to a binary that can be consumed directly by OpenCV/Evision. The `from_binary/5` function in Evision requires the dimensions of the image, and using the binary extracted from erl_tar gives the same size mismatch error as before.

Example:

```ex
iex> bin_images |> List.first() |> OpenCV.Mat.from_binary({:u, 8}, 200, 200, 3)
{:error, 'size mismatch'}
```

josevalim commented:
So we have a function in Evision/OpenCV to open an image from memory/binary instead of from disk?

goodhamgupta commented:
> So we have a function in Evision/OpenCV to open an image from memory/binary instead of from disk?

Indeed! From the evision documentation:

```ex
## to_binary/from_binary
iex> {:ok, mat} = OpenCV.imread("/path/to/img.jpg")
iex> {:ok, type} = OpenCV.Mat.type(mat)
iex> {:ok, {rows, cols, channels}} = OpenCV.Mat.shape(mat)
iex> {:ok, binary_data} = OpenCV.Mat.to_binary(mat)
 # read from binary
iex> OpenCV.Mat.from_binary(binary_data, type, cols, rows, channels)
```
Note that this will not work with the binary obtained from erl_tar, because of the size mismatch.

josevalim commented:
Sorry, I think I did not express myself correctly or maybe I am misunderstanding it. We have two binaries: the file in bytes and the matrix representation. We need an API that receives the file in bytes and returns the `mat`.

goodhamgupta commented:
> Sorry, I think I did not express myself correctly or maybe I am misunderstanding it. We have two binaries: the file in bytes and the matrix representation. We need an API that receives the file in bytes and returns the `mat`.

This makes sense. Sorry, I misunderstood! 

I'm not aware of how to convert this binary to a `mat`. Looking through the evision code, would it be similar to implementing the following line: https://github.com/cocoa-xu/evision/blob/main/c_src/modules/opencv_mat.h#L113 ?

josevalim commented:
Maybe @cocoa-xu can help us? @cocoa-xu, is it possible to open up an image (jpg/png) but, instead of giving it a filename, we give the file contents already loaded in memory?

cocoa-xu commented on lib/scidata/caltech101.ex:
```diff
@@ -0,0 +1,83 @@
+    records =
+      files
+      |> Task.async_stream(&generate_records(&1, label_mapping, image_mapping),
+        max_concurrency: Application.get_env(:scidata, :max_concurrency)
```
Perhaps set a default value for `Application.get_env`? e.g., 
```
Application.get_env(:scidata, :max_concurrency, 1)
```

Otherwise, `max_concurrency` will be `nil` which is invalid for `Task.async_stream`.

josevalim replied:
We should use the default, which is based on the number of machine cores, but allow an `opts` argument to be given which may change it?

cocoa-xu commented on lib/scidata/caltech101.ex:
```diff
@@ -0,0 +1,83 @@
+  end
+
+  defp get_mapping(:label) do
+    fpath = Path.join([File.cwd!(), "lib/scidata/labels/caltech101_labels.json"])
```
These assets/resource files should be put in the `./priv` directory, and use `:code.priv_dir(:scidata)` instead of `File.cwd!()`.

It should look like the following.

```
.
├── LICENSE
├── README.md
├── lib
│   └── scidata
│       ├── caltech101.ex
│       ├── cifar10.ex
│       ├── cifar100.ex
│       ├── content
│       │   └── caltech101_image_shapes.json
│       ├── fashionmnist.ex
│       ├── imdb_reviews.ex
│       ├── kuzushiji_mnist.ex
│       ├── labels
│       │   └── caltech101_labels.json
│       ├── mnist.ex
│       ├── squad.ex
│       ├── utils.ex
│       ├── yelp_full_reviews.ex
│       └── yelp_polarity_reviews.ex
├── mix.exs
├── mix.lock
├── priv
│   ├── content
│   │   └── caltech101_image_shapes.json
│   └── labels
│       └── caltech101_labels.json
└── test
    ├── caltech101_test.exs
    ├── imdb_reviews_test.exs
    ├── kuzushiji_mnist_test.exs
    ├── mnist_test.exs
    ├── scidata_test.exs
    ├── squad_test.exs
    ├── test_helper.exs
    ├── yelp_full_reviews_test.exs
    └── yelp_polarity_reviews_test.exs
```

cocoa-xu replied:
And these two functions should be

```elixir
  defp get_mapping(:label) do
    fpath = Path.join([:code.priv_dir(:scidata), "labels/caltech101_labels.json"])
    do_get_mapping(fpath)
  end

  defp get_mapping(:image) do
    fpath = Path.join([:code.priv_dir(:scidata), "content/caltech101_image_shapes.json"])
    do_get_mapping(fpath)
  end
```

cocoa-xu reviewed (changes requested):
I'm still trying to run the download part, but I think decoding an image loaded in memory is totally achievable.

cocoa-xu commented:
Now it should be okay with evision [HEAD](https://github.com/cocoa-xu/evision/tree/1592b1127c42be590c5089376c299ab0f795f29d).

```
iex> {images, labels} = Scidata.Caltech101.download()
iex> {bin_images, _img_type, img_shapes} = images
iex> {:ok, mat} = bin_images |> List.first |> OpenCV.imdecode(OpenCV.cv_imread_color)
iex> {:ok, shape} = OpenCV.Mat.shape(mat)
iex> ^shape = img_shapes |> List.first
```

cocoa-xu commented:
Also, I think the file `lib/scidata/content/caltech101_image_shapes.json` is unnecessary. Because we will need to decode jpg/png files, and `evision` or any other image decoders should decode them correctly and return the shape information as long as the file is not corrupted.

cocoa-xu commented:
But if we would like to avoid using any image decoders, maybe we can manually convert this dataset into a single binary file and store it somewhere online as this dataset barely changes, and then we can download the binary file and parse it recursively in `scidata`

caltech101_helper.ex
```elixir
defmodule Caltech101.Helper do
  def to_binary(filename) when is_binary(filename) do
    {images, {labels, _}} = Scidata.Caltech101.download()
    {bin_images, img_shapes} = images
    labels = :binary.bin_to_list(labels)
    binary = for index <- 0..(Enum.count(bin_images) - 1), reduce: <<>> do
      acc ->
        {h, w, c} = Enum.at(img_shapes, index)
        class = Enum.at(labels, index)
        bin = Enum.at(bin_images, index)
        acc <> <<h::32-little, w::32-little, c::8, class::8>> <> bin
    end
    {:ok, file} = File.open(filename, [:write])
    IO.binwrite(file, binary)
    File.close(file)
    binary
  end

  def from_binary(filename) when is_binary(filename) do
    {:ok, dataset} = File.read(filename)
    parse_image(dataset)
  end

  defp parse_image(binary), do: parse_image(binary, [], [], [])
  defp parse_image(<<>>, images, labels, shapes), do: {images, labels, shapes}
  defp parse_image(<<h::32-little, w::32-little, c::8, class::8, restData::binary>>, images, labels, shapes) do
    bin_size = w * h * c
    image = :binary.part(restData, 0, bin_size)
    restImage = :binary.part(restData, bin_size, byte_size(restData) - bin_size)
    images = [image | images]
    labels = [class | labels]
    shapes = [{h, w, c} | shapes]
    parse_image(restImage, images, labels, shapes)
  end
end
```

caltech101.ex
```elixir
defmodule Scidata.Caltech101 do
  @moduledoc """
  Module for downloading the [Caltech101 dataset](http://www.vision.caltech.edu/Image_Datasets/Caltech101).
  """

  require Scidata.Utils
  alias Scidata.Utils
  require OpenCV

  # NOTE: The original url on the website is "https://drive.google.com/u/0/uc?export=download&confirm=R4MY&id=137RyRjvTBkBiIfeYBNZBtViDHQ6_Ewsp".
  # However, I was unable to get around the redirection issue when downloading large files from GDrive.
  @base_url "https://s3.amazonaws.com/fast-ai-imageclas/caltech_101.tgz"
  @labels_shape {9144, 1}

  @doc """
  Downloads the Caltech101 training dataset or fetches it locally.

  Returns a tuple of format:

      {{images_binary, images_type, images_shape},
       {labels_binary, labels_type, labels_shape}}

  If you want to one-hot encode the labels, you can:

      labels_binary
      |> Nx.from_binary(labels_type)
      |> Nx.new_axis(-1)
      |> Nx.equal(Nx.tensor(Enum.to_list(1..102)))

  """
  def download() do
    download_dataset(:train)
  end

  defp download_dataset(_dataset_type) do
    # Skip first file since it's a temporary file.
    [_ | files] = Utils.get!(@base_url).body

    {:ok, label_mapping} = get_mapping(:label)

    records =
      files
      |> Task.async_stream(&generate_records(&1, label_mapping),
        max_concurrency: Application.get_env(:scidata, :max_concurrency, 1)
      )
      |> Enum.to_list()

    images = Enum.map(records, fn {:ok, record} -> elem(record, 0) end)
    labels = Enum.map(records, fn {:ok, record} -> elem(record, 1) end)
    shapes = Enum.map(records, fn {:ok, record} -> elem(record, 2) end)

    {{images, shapes}, {IO.iodata_to_binary(labels), @labels_shape}}
  end

  defp get_mapping(:label) do
    fpath = Path.join([:code.priv_dir(:scidata), "labels/caltech101_labels.json"])
    do_get_mapping(fpath)
  end

  defp do_get_mapping(fpath) do
    with {:ok, body} <- File.read(fpath),
         {:ok, mapping} <- Jason.decode(body),
         do: {:ok, mapping}
  end

  defp generate_records({fname, image}, label_mapping) do
    [_, class_name, image_fname] =
      fname
      |> List.to_string()
      |> String.downcase()
      |> String.split("/")

    label = Map.get(label_mapping, class_name)

    {:ok, image_mat} = OpenCV.imdecode(image, OpenCV.cv_imread_color())
    {:ok, binary} = OpenCV.Mat.to_binary(image_mat)
    {:ok, shape} = OpenCV.Mat.shape(image_mat)
    {binary, label, shape}
  end
end
```

The result binary file size will be ~2GB and `zip -9` gives a 1.25GB file.

cocoa-xu commented:
For the image decoding task, I just wrote a tiny library, [img_decode](https://github.com/cocoa-xu/img_decode) based on [stb_image](https://github.com/nothings/stb/blob/master/stb_image.h). Though the number of supported image formats in `stb_image` is less than OpenCV, it's good enough for most public image datasets. Adding `evision` as a dependency will have a larger footprint (even when only `imgcodecs`, `imgproc` and `core` modules are enabled).

josevalim commented:
Thank you @cocoa-xu, this is amazing! There is one typo in the project description I think: ting -> tiny. :)

goodhamgupta commented:
> For the image decoding task, I just wrote a tiny library, img_decode based on stb_image. Though the number of supported image formats in stb_image is less than OpenCV, it's good enough for most public image datasets. Adding evision as a dependency will have a larger footprint (even when only imgcodecs, imgproc and core modules are enabled).

Wow thank you so much for the library @cocoa-xu! 😄 I've replaced evision with `img_decode` and addressed all your previous comments.

Note that, when I try to compile `img_decode` on my mac(not M1), I get the following error:

```
Erlang/OTP 24 [erts-12.0] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [jit]

==> img_decode
-- The C compiler identification is AppleClang 12.0.5.12050022
-- The CXX compiler identification is AppleClang 12.0.5.12050022
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    PRIV_DIR


-- Build files have been written to: /Users/Shubham/shubham/scidata/_build/dev/lib/img_decode/cmake_imgdecode
[ 50%] Building CXX object CMakeFiles/imgdecode.dir/c_src/imgdecode_stb.cpp.o
clang: warning: argument unused during compilation: '-undefined dynamic_lookup' [-Wunused-command-line-argument]
/Users/Shubham/shubham/scidata/deps/img_decode/c_src/imgdecode_stb.cpp:3:10: error: 'stb_image.h' file not found with <angled> include; use "quotes" instead
#include <stb_image.h>
         ^~~~~~~~~~~~~
         "stb_image.h"
1 error generated.
make[3]: *** [CMakeFiles/imgdecode.dir/c_src/imgdecode_stb.cpp.o] Error 1
make[2]: *** [CMakeFiles/imgdecode.dir/all] Error 2
make[1]: *** [all] Error 2
make: *** [/Users/Shubham/shubham/scidata/_build/dev/lib/img_decode/priv/imgdecode.so] Error 2
could not compile dependency :img_decode, "mix compile" failed. You can recompile this dependency with "mix deps.compile img_decode", update it with "mix deps.update img_decode" or clean it with "mix deps.clean img_decode"
==> scidata
** (Mix) Could not compile with "make" (exit status: 2).
You need to have gcc and make installed. Try running the
commands "gcc --version" and / or "make --version". If these programs
are not installed, you will be prompted to install them.
```

I guess I'm missing some package, but  I don't know which one. For now, the dirty hack I've used to fix this is:

- Download the `std_image.h` file from the `std` repository and place it in the same repository as the `imgdecode_stb.cpp`( in `img_decode/c_src`)
- Replace the angular brackets on the line(https://github.com/cocoa-xu/img_decode/blob/main/c_src/imgdecode_stb.cpp#L3) with quotes i.e use:
```cpp
#include "stb_image.h"
```

josevalim commented:
If stb_image does not use cmake itself, perhaps we can skip cmake altogether and use a regular makefile? I can try to send a PR too.

cocoa-xu commented:
Hi @goodhamgupta, `stb_image.h` should be located in `3rd_party/stb`, which is a git submodule. I'll add `git submodule update --init --recursive` to the Makefile in a second!

cocoa-xu commented:
I've updated it just now, `mix deps.update img_decode` should be able to make it point to the latest commit.

goodhamgupta commented:
Thanks @cocoa-xu ! 😄  I've updated the dependency and can confirm that it works as expected!

cocoa-xu commented:
> If stb_image does not use cmake itself, perhaps we can skip cmake altogether and use a regular makefile? I can try to send a PR too.

I agree with you, for current status, we don't really need cmake. I was thinking about adding the webp decoder as `stb_image` does not support that, which has a fairly good chance to appear in future datasets to reduce the file size. 

But it's probably already good enough for quite a while. I'll try to only use a regular Makefile so that we don't have to rely on cmake.

cocoa-xu commented:
Okay, img_decode is cmake free now! @goodhamgupta could you please update the dependency one more time and check if it still works now? Thanks in advance!

goodhamgupta commented:
> Okay, img_decode is cmake free now! @goodhamgupta could you please update the dependency one more time and check if it still works now? Thanks in advance!

This is done and I can confirm that everything still works as expected! 😄

josevalim commented on config/config.exs:
```diff
@@ -0,0 +1,4 @@
+import Config
+
+config :scidata, 
+  max_concurrency: "MAX_CONCURRENCY" |> System.get_env("#{System.schedulers_online}") |> String.to_integer
```
Note this configuration only affects scidata itelf. It is not used by actual projects. In any case, avoiding global configuration is recommended, so we should prefer options, as I left a note below. :)

josevalim commented on priv/labels/caltech101_labels.json:
```diff
@@ -0,0 +1,104 @@
+   "windsor_chair": 99,
+   "wrench": 100,
+   "yin_yang": 101
+}
```
Let's convert this to a map and embed it in Elixir directly. Less work to do when we actually get to download stuff :)

goodhamgupta commented:
> We should use the default, which is based on the number of machine cores, but allow an opts argument to be given which may change it?

Good idea! 😄  I've made this change as well.

josevalim commented on lib/scidata/caltech101.ex:
```diff
@@ -0,0 +1,170 @@
+    records =
+      files
+      |> Task.async_stream(&generate_records/1,
+        max_concurrency: Keyword.get(opts, :max_concurrency, System.schedulers_online)
```
```suggestion
        max_concurrency: Keyword.get(opts, :max_concurrency, System.schedulers_online() * 2),
        timeout: :infinity
```

I doubled the amount of cores since there is some IO bound work. :)

josevalim commented on lib/scidata/caltech101.ex:
```diff
@@ -0,0 +1,170 @@
+      |> Enum.at(1)
+      |> String.to_atom()
+
+    label = Map.get(@label_mapping, class_name)
```
```suggestion
    label = Map.fetch!(@label_mapping, class_name)
```

The above will raise if the label does not exist in the mapping, WDYT?

josevalim commented on mix.exs:
```diff
@@ -32,7 +32,8 @@ defmodule Scidata.MixProject do
       {:nimble_csv, "~> 1.1"},
-      {:jason, "~> 1.0"}
+      {:jason, "~> 1.0"},
+      {:img_decode, "~> 0.1.0", github: "cocoa-xu/img_decode"}
```
Final suggestion. Let's make this an optional dependency:

```suggestion
      {:img_decode, "~> 0.1.0", github: "cocoa-xu/img_decode", optional: true}
```

Then at the beginning of the download function you check this:

```elixir
unless Code.ensure_loaded?(ImgDecode) do
  raise "ImgDecode is missing, please add `{:img_decode, "~> 0.1"}` as a dependency to your mix.exs"
end
```

goodhamgupta commented:
> The above will raise if the label does not exist in the mapping, WDYT?

Ah this would have caused `nil` labels initially. Even though the dataset is unlikely to change in the future, I agree that we should replace `get` with `fetch!`.

cocoa-xu commented on mix.exs:
```diff
@@ -32,7 +32,8 @@ defmodule Scidata.MixProject do
       {:nimble_csv, "~> 1.1"},
-      {:jason, "~> 1.0"}
+      {:jason, "~> 1.0"},
+      {:img_decode, "~> 0.1.0", github: "cocoa-xu/img_decode", optional: true}
```
```suggestion
      {:stb_image, "~> 0.1.0", github: "cocoa-xu/stb_image", optional: true}
```
Hi, I've made some changes in the upstream and the name of this library has changed to `stb_image`.

cocoa-xu commented on lib/scidata/caltech101.ex:
```diff
@@ -0,0 +1,173 @@
+      |> String.to_atom()
+
+    label = Map.fetch!(@label_mapping, class_name)
+    {:ok, image_bin, image_shape, _img_type} = ImgDecode.from_memory(image)
```
```suggestion
    {:ok, image_bin, image_shape, _img_type} = StbImage.from_memory(image)
```

cocoa-xu commented on lib/scidata/caltech101.ex:
```diff
@@ -0,0 +1,173 @@
+
+  """
+  def download(opts \\ []) do
+    unless Code.ensure_loaded?(ImgDecode) do
+      raise "ImgDecode is missing, please add `{:img_decode, \"~> 0.1\"}` as a dependency to your mix.exs"
```
```suggestion
    unless Code.ensure_loaded?(StbImage) do
      raise "StbImage is missing, please add `{:stb_image, \"~> 0.1\"}` as a dependency to your mix.exs"
```

josevalim reviewed (commented):
Hi @goodhamgupta! This looks almost ready to go, just some final book keeping left after the merge. Please do this:

1. Pull latest changes in your machine
2. Run `mix deps.get` so the lockfile is updated
3. Run `mix deps.unlock --unused` to remove any unused dep from the lock (such as `img_decode`)
4. Run `mix format`

Then we should be all set!

goodhamgupta commented:
Thank you for the detailed steps! 😄 My apologies for the continuous back-and-forth after every change 😅

josevalim commented:
It is completely fine and expected, no worries at all!

t-rutten commented on lib/scidata/caltech101.ex:
```diff
@@ -0,0 +1,174 @@
+      |> String.to_atom()
+
+    label = Map.fetch!(@label_mapping, class_name)
+    {:ok, image_bin, image_shape, _img_type} = StbImage.from_memory(image)
```
Running `download/1` locally, this match is failing for e.g. '101_ObjectCategories/BACKGROUND_Google/image_0004.jpg' because `StbImage.from_memory(image)` is returning
```
{:ok, <<44, 49, 26, ...>>, {200, 200, 3}, :u8, :rgb}
```
and we're not expecting the last element of the return tuple.

Maybe I'm pulling the wrong version of `:stb_image`? Or the API has changed?

cocoa-xu replied:
```suggestion
    {:ok, image_bin, image_shape, _img_type, _img_channels} = StbImage.from_memory(image)
```

Oh, sorry that I missed this part. I added another element (`image_channels`) in the return tuple a few days ago, and I forgot to suggest the corresponding changes here.

t-rutten replied:
No worries! Thanks for confirming here.

t-rutten commented on lib/scidata/caltech101.ex:
```diff
@@ -0,0 +1,174 @@
+  require Scidata.Utils
+  alias Scidata.Utils
+
+  # NOTE: The original url on the website is "https://drive.google.com/u/0/uc?export=download&confirm=R4MY&id=137RyRjvTBkBiIfeYBNZBtViDHQ6_Ewsp".
+  # However, I was unable to get around the redirection issue when downloading large files from GDrive.
```
```suggestion
```
We have this documented in the PR discussion :)

goodhamgupta commented:
@t-rutten Thanks for your review! The pattern match was indeed incorrect. I don't remember seeing this change in the API, but I must've missed it! 😅 I've made the required changes now.

cocoa-xu commented:
@goodhamgupta sorry I forgot to suggest this change. I was looking for replacing all `img_decode` with `stb_image` and then I forgot this.😅

goodhamgupta commented:
> @goodhamgupta sorry I forgot to suggest this change. I was looking for replacing all img_decode with stb_image and then I forgot this.😅

No worries! 😄  I admittedly forgot to re-run the tests after making the changes, which should've helped catch this change earlier as well! 😅

josevalim commented on mix.exs:
```diff
@@ -32,7 +32,8 @@ defmodule Scidata.MixProject do
       {:nimble_csv, "~> 1.1"},
-      {:jason, "~> 1.0"}
+      {:jason, "~> 1.0"},
+      {:stb_image, "~> 0.1.0", github: "cocoa-xu/stb_image", optional: true}
```
Just a heads up: we will need to convert it to a package before a package release (as packages cannot have git deps).

t-rutten replied:
Thanks for the reminder @josevalim!

@cocoa-xu would you be willing to publish `stb_img` on hex.pm?

cocoa-xu replied:
Of course! I just published it is on hex.pm.

As a side note, it seemed that publishing to hex.pm will remove the git submodule info (not sure which files are necessary to keep the submodule info, and I'm not sure if I should go with the submodule way when publishing to hex.pm). Therefore, I downloaded the `stb_image.h` and its LICENSE file and gave up using git submodule.

```suggestion
      {:stb_image, "~> 0.1.2", optional: true}
```

I tested this version on my machine, and it worked as expected.

t-rutten replied:
Thank you @cocoa-xu for your critical help here. :)

cocoa-xu replied:
Not at all! I'm glad I could be of help! 😄
